splice_machine/src/main/java/com/splicemachine/derby/stream/function/merge/AbstractMergeJoinFlatMapFunction.java

```diff
@@ -119,49 +119,65 @@ protected void initRightScan(PeekingIterator<ExecRow> leftPeekingIterator) throw
+
+        // this is 0-based array, the column positions in the array are also 0-based.
+        int[] colToBaseTableMap = ((MergeJoinOperation)joinOperation).getRightHashKeyToBaseTableMap();
+
+        // we cannot map the hash fields back to the base table column position, so cannot utilize the hash field value
+        // to restrict the rigth table scan
+        if (colToBaseTableMap == null)
             scanStartOverride = startPosition;
-        } else {
-            // this is 0-based array, the column positions in the array are also 0-based.
-            int[] colToBaseTableMap = ((MergeJoinOperation)joinOperation).getRightHashKeyToBaseTableMap();
-
-            // we cannot map the hash fields back to the base table column position, so cannot utilize the hash field value
-            // to restrict the rigth table scan
-            if (colToBaseTableMap == null)
-                scanStartOverride = startPosition;
-            else {
-
-                ExecRow newStartKey = new ValueRow(columnOrdering.length);
-                int[] rightHashKeys = joinOperation.getRightHashKeys();
-
-                for (int i = 0; i < nCols; i++)
-                    newStartKey.setColumn(i + 1, startPosition.getColumn(i + 1));
-
-                int len = nCols;
-                for (int i = len; i < columnOrdering.length; i++) {
-                    int keyColumnBaseTablePosition = columnOrdering[i];
-                    int j = 0;
-                    for (; j < rightHashKeys.length; j++) {
-                        // both columnOrdering and colToBaseTableMap are 0-based
-                        if (colToBaseTableMap[j] == keyColumnBaseTablePosition) {
+        else {
+
+            ExecRow newStartKey = new ValueRow(columnOrdering.length);
+            int[] rightHashKeys = joinOperation.getRightHashKeys();
+            int[] rightHashKeySortOrders = ((MergeJoinOperation) joinOperation).getRightHashKeySortOrders();
+
+            int len = 0;
+            for (int i = 0; i < columnOrdering.length; i++) {
+                int keyColumnBaseTablePosition = columnOrdering[i];
+                int j = 0;
+                for (; j < rightHashKeys.length; j++) {
+                    // both columnOrdering and colToBaseTableMap are 0-based
+                    if (colToBaseTableMap[j] == keyColumnBaseTablePosition) {
+                        if (i >= nCols) {
                             newStartKey.setColumn(i + 1, firstHashRow.getColumn(j + 1));
-                            len++;
-                            break;
                         }
-                    }
-                    //no match found for the given key column position
-                    if (j >= rightHashKeys.length)
+                        else {
+                            // If a descending index, reverse the result of the comparison.
+                            boolean replaceWithLeftRowVal = (startPosition == null) ||
```

Comment:
This logic is a bit hard to follow, could you break it down into one or more utility methods or at least use a few more booleans if possible?